### ISSUE TITLE
S1118 on abstract classes

### ISSUE THREAD

**Issue by splendid25** [SonarSource/sonar-csharp#1411]
### Description
Fix the S1118 recommendation on abstract classes.

### Repro steps
```
public abstract class MyAbstractClass
{
    public static void Action() { }
}
```

### Expected behavior

No warning since the default constructor can only be called from derived classes.
Or just the second part :
Add the 'static' keyword to the class declaration.

What do you think?

### Actual behavior
Warning	S1118	Add a 'protected' constructor or the 'static' keyword to the class declaration.

### Known workarounds
Following the recommendation https://rules.sonarsource.com/csharp/RSPEC-1118 :
1. It just adds some noise, but not relevant :
```
public abstract class MyAbstractClass
{
    protected MyAbstractClass() { }
    public static void Action() { }
}
```

2. Might be better, but is a breaking change when there are already derived classes :
```
public static class MyAbstractClass
{
    public static void Action() { }
}
```

### Related information
* SonarC# Version 4.0.0.3479
* Visual Studio Version 15.7.1

**Comment by agigleux** [comment]
Hello @splendid25,

I agree with you we should not raise an issue for S1118 on abstract classes because there is no way to call a static method on a class instance. Adding an extra "protected" constructor is not helping for readability.

**Comment by agigleux** [comment]
https://jira.sonarsource.com/browse/RSPEC-3115 should be updated accordingly once fix is implemented.

**Comment by piyushparashar** [comment]
Hi,

Should this rule be reported on sealed classes?
